src/main/java/co/cask/publisher/GooglePublisher.java

```diff
@@ -0,0 +1,151 @@
+public class GooglePublisher extends BatchSink<StructuredRecord, NullWritable, Text> {
+  private final Config config;
+
+  @SuppressWarnings("unused")
+  public GooglePublisher(Config config) {
+    this.config = config;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    super.configurePipeline(pipelineConfigurer);
+    config.validate();
+  }
+
+  @Override
+  public void prepareRun(BatchSinkContext context) throws Exception {
+    config.validate();
+    Configuration configuration = new Configuration();
+    PubSubOutputFormat.configure(configuration, config);
+    context.addOutput(Output.of(config.referenceName,
+                                new SinkOutputFormatProvider(PubSubOutputFormat.class, configuration)));
+  }
+
+  @Override
+  public void transform(StructuredRecord input, Emitter<KeyValue<NullWritable, Text>> emitter) throws Exception {
+    String body = StructuredRecordStringConverter.toJsonString(input);
+    emitter.emit(new KeyValue<>(NullWritable.get(), new Text(body)));
+  }
+
+  /**
+   * PubSub Publisher config
+   */
+  public static class Config extends GCPReferenceConfig {
+    @Description("Cloud Pub/Sub topic to publish the records")
+    @Macro
+    public String topic;
+
+    // batching options
+    @Description("Element count threshold used for batching, default : 100")
+    @Macro
+    @Nullable
+    public Long messageCountBatchSize;
+
+    @Description("Request byte threshold used for batching, default : 1kb")
+    @Macro
+    @Nullable
+    public Long requestThresholdKB;
+
+    @Description("Delay threshold in millis from last publish for publishing the next batch, default : 1 ms")
+    @Macro
+    @Nullable
+    public Long publishDelayThresholdMillis;
+
+    @Description("Maximum number of publish message errors to tolerate per mapper before throwing an error")
+    @Macro
+    @Nullable
+    public Long errorThreshold;
+
+    @Description("Max retry timeout on errors while publishing messages to pubsub, default value is 30 seconds")
+    @Macro
+    @Nullable
+    public Integer retryTimeoutSeconds;
+
+
+    public void validate() {
+      if (!containsMacro("messageCountBatchSize") && messageCountBatchSize != null && messageCountBatchSize < 1) {
+        throw new IllegalArgumentException("Maximum count of messages in a batch should be positive for Pub/Sub");
+      }
+      if (!containsMacro("requestThresholdKB") && requestThresholdKB != null && requestThresholdKB < 1) {
+        throw new IllegalArgumentException("Maximum size of a batch (KB) should be positive for Pub/Sub");
+      }
+      if (!containsMacro("publishDelayThresholdMillis") &&
+        publishDelayThresholdMillis != null && publishDelayThresholdMillis < 1) {
+        throw new IllegalArgumentException("Delay threshold for publishing a batch should be positive for Pub/Sub");
+      }
+      if (!containsMacro("errorThreshold") && errorThreshold != null && errorThreshold < 1) {
```

Comment:
shouldn't an error threshold of 0 be allowed? It would mean a single error fails the job.

Reply:
yes, it should be allowed, good catch, updated it.